client/src/util/stateManager/matrix.ts

```diff
@@ -12,30 +12,68 @@ import {
   KeyIndex,
 } from "../dataframe";
 
+import { CatInt8Array } from "./cat8_array";
+import { CatInt16Array } from "./cat16_array";
+import { CatInt32Array } from "./cat32_array";
+
 const utf8Decoder = new TextDecoder("utf-8");
 
 /**
  * Matrix flatbuffer decoding support. See fbs/matrix.fbs
  */
 
 /**
- * Decode NetEncoding.TypedArray
+ * Decode NetEncoding.TypedFBArray
  */
 // eslint-disable-next-line @typescript-eslint/no-explicit-any --- FIXME: disabled temporarily on migrate to TS.
-function decodeTypedArray(uType: any, uValF: any, inplace = false) {
-  if (uType === NetEncoding.TypedArray.NONE) {
+function decodeTypedArray(uType: any, uValF: any, inplace = false): any {
+  if (uType === NetEncoding.TypedFBArray.NONE) {
     return null;
   }
 
   // Convert to a JS class that supports this type
   // @ts-expect-error --- FIXME: Element implicitly has an 'any' type.
-  const TypeClass = NetEncoding[NetEncoding.TypedArray[uType]];
+  const TypeClass = NetEncoding[NetEncoding.TypedFBArray[uType]];
   // Create a TypedArray that references the underlying buffer
-  let arr = uValF(new TypeClass()).dataArray();
-  if (uType === NetEncoding.TypedArray.JSONEncodedArray) {
+  const x = uValF(new TypeClass());
+  let arr = x.dataArray();
+  let dict;
+  let flag = true;
+  if (uType === NetEncoding.TypedFBArray.JSONEncodedFBArray) {
     const json = utf8Decoder.decode(arr);
     arr = JSON.parse(json);
-  } else if (!inplace) {
+    flag = false;
+  } else if (
+    uType === NetEncoding.TypedFBArray.CatInt8FBArray ||
+    uType === NetEncoding.TypedFBArray.CatInt16FBArray ||
+    uType === NetEncoding.TypedFBArray.CatInt32FBArray
+  ) {
+    dict = JSON.parse(utf8Decoder.decode(x.codesArray()));
+    if (dict) {
+      if (uType === NetEncoding.TypedFBArray.CatInt8FBArray) {
+        arr = new CatInt8Array({ array: arr, codeMapping: dict });
+      } else if (uType === NetEncoding.TypedFBArray.CatInt16FBArray) {
+        arr = new CatInt16Array({ array: arr, codeMapping: dict });
+      } else if (uType === NetEncoding.TypedFBArray.CatInt32FBArray) {
+        arr = new CatInt32Array({ array: arr, codeMapping: dict });
+      }
+    }
+    flag = false;
+  } else if (
+    uType === NetEncoding.TypedFBArray.SparseFloat32FBArray ||
+    uType === NetEncoding.TypedFBArray.SparseFloat64FBArray
+  ) {
+    const rows = x.rowsArray();
+    const size = x.sizeArray()[0];
+    const data = new arr.constructor(size);
+    for (const [i, row] of rows.entries()) {
```

Comment:
is there a reason you use for-of instead of a simple for loop?  My experience is that iterators have traditionally been slower, and this is a spot you _really_ care about perf as it is a giant memory copy.

Ie,
```
for (const i = 0, nrows = rows.length; i < nrows; i += 1) {
    data[rows[i]] = arr[i];
}
```

If you do have a preference for for-of, I would do some benchmarking to make sure that it isn't (still) slower.